daemon/cmd/daemon.go

```diff
@@ -582,8 +582,76 @@ func NewDaemon(ctx context.Context, cancel context.CancelFunc, epMgr *endpointma
 	debug.RegisterStatusObject("ongoing-endpoint-creations", d.endpointCreations)
 
 	d.k8sWatcher.RunK8sServiceHandler()
+
+	if option.Config.DNSPolicyUnloadOnShutdown {
+		log.Debug("Registering cleanup function to unload DNS policies")
+
+		// add to pre-cleanup funcs because this needs to run on graceful shutdown, but
+		// before the relevant subystems are being shut down.
+		cleaner.preCleanupFuncs.Add(func() {
+			// Stop k8s watchers
+			log.Info("Stopping k8s service handler")
+			d.k8sWatcher.StopK8sServiceHandler()
+
+			// Iterate over the policy repository and remove L7 DNS part
+			removeL7DNSRules := func(pr policyAPI.Ports) error {
+				portProtocols := pr.GetPortProtocols()
+				if len(portProtocols) != 1 || portProtocols[0].Port != "53" {
+					// not a L7 DNS rule
+					return nil
+				}
+				portRule := pr.GetPortRule()
+				if portRule == nil || portRule.Rules == nil {
+					// no L7 rule
+					return nil
+				}
+				l7Rules := portRule.Rules
+				log.Debugf("Found egress L7 DNS rule, portProtocol %#v l7 rules %#v", portProtocols[0], *l7Rules)
+
+				// For security reasons, the L7 DNS policy must be a
+				// wildcard in order to trigger this logic.
+				// Otherwise we could invalidate the L7 security
+				// rules. This means if any of the DNS L7 rules
+				// have a matchPattern of * then it is OK to delete
+				// the L7 portion of those rules.
+				hasWildcard := false
+				for _, dns := range l7Rules.DNS {
+					if dns.MatchPattern == "*" {
+						hasWildcard = true
+						break
+					}
+				}
+				if hasWildcard {
+					portRule.Rules = nil
+				}
+				return nil
+			}
+
+			policyRepo := d.GetPolicyRepository()
+			policyRepo.Mutex.Lock()
+			policyRepo.IterateLocked(func(rule *policyAPI.Rule) {
+				for _, er := range rule.Egress {
+					_ = er.ToPorts.Iterate(removeL7DNSRules)
+				}
+				// TODO: modify Ingress, IngressDeny and EgressDeny as well?
```

Comment:
@joestringer do we need to modify ingress? I don't think so.

@tklauser regarding the [Ingress|Egress]Deny, they should be left untouched.

Reply:
Pretty sure we have a docs page somewhere that says that DNS only supports egress. Agree we shouldn't need to worry there. +1 on not touching Denies.

Reply:
Thanks for clarifying, I've removed the comment.